### PR TITLE
ELF64 support

### PR DESCRIPTION
**Pull request by masayuki2009**
### Summary

- This PR adds ELF64 support to NuttX.
- To use ELF64, CONFIG_ELF_64BIT must be enabled.
- To support ELF64, some common definitions have been moved from elf32.h to elf_common.h
- Also, Elf_xxx have been introduced to be used in common libraries such as binfmt.
- However, Elf32_xxx are still valid to be mainly used in arch_elf.c (32bit CPU only)
- Add RISC-V 64bit ELF support

### Impact

- This PR affects common libraries such as binfmt but should work

### Limitations / TODO

- Currently 32bit ELF for RISC-V is not supported

### Testing

-  Tested with following environments
- (32bit cpu) spresense:elf, spresense:posix_spawn, spresense:module 
- (64bit cpu) maix-bit:elf, maix-bit:posix_spawn

### CODE REVIEW

**Comment by masayuki2009** [comment]
Please note that some typedefs such as Elf32_xxx, Elf64_xxx, Elf_xxx still violate NuttX conding standard. Because modifying all such typedefs impacts many files, so in this PR, I focused on Elf64 support with minimal changes.

File: binfmt/libelf/libelf_bind.c [hunk 1]
```diff
@@ -326,8 +367,181 @@ static int elf_relocate(FAR struct elf_loadinfo_s *loadinfo, int relidx,
 static int elf_relocateadd(FAR struct elf_loadinfo_s *loadinfo, int relidx,
```
Comment (xiaoxiang781216) [inline binfmt/libelf/libelf_bind.c]: Let's enhance modlib_relocateadd too, maybe it's better to split RELA to another patch.

File: include/elf_common.h [hunk 2]
```diff
@@ -0,0 +1,214 @@
+/****************************************************************************
+ * include/elf_common.h
```
[the remark this answers is not in the document]
Comment (masayuki2009) [reply include/elf_common.h]: > It is better to rename elf_common.h to elf.h to more compatible with other POSIX variant, 
> and include elf32.h/elf64.h inside elf.h like this:

@xiaoxiang781216  Thanks for the comment. I'll modify the file later.

File: binfmt/libelf/libelf_bind.c (hunk 1 above)
Comment (masayuki2009) [reply binfmt/libelf/libelf_bind.c]: > Let's enhance modlib_relocateadd too, 
> maybe it's better to split RELA to another patch.

@xiaoxiang781216 Thanks for the comments. I'll consider them later.

**Comment by patacongo** [comment]
"Please note that some typedefs such as Elf32_xxx, Elf64_xxx, Elf_xxx still violate NuttX conding standard."  Those are the names used in other elf header files.  I have missed feelings if they should be made to match the coding standard or not.  They are controlled by any standard, so the answer is that they should be made to follow the NuttX coding standard someday.

Given that Masayuki has suggested waiting for subsequent PRs to address the comments, is this PR ready to be merged now?

**Comment by masayuki2009** [comment]
@patacongo Thanks for the comment regarding coding standard. Now I'm preparing new patch sets which @xiaoxiang781216 pointed out. Perhaps, I need one more day...

**Comment by xiaoxiang781216** [comment]
> someday

But this is a defacto standard, should we follow the well established convention?

**Comment by patacongo** [comment]
Yes, but it is not a standard.

I don't know what the answer is. I understand what you are saying, but if we ignore the coding standard, then I think we are also asking for problems in the future.  I think it is safer to just always following the coding standard unless there is an established standard that conflicts with it.

**Comment by xiaoxiang781216** [comment]
> Yes, but it is not a standard.
> 
> I don't know what the answer is. I understand what you are saying, but if we ignore the coding standard, then I think we are also asking for problems in the future. I think it is safer to just always following the coding standard unless there is an established standard that conflicts with it.

Why we follow the standard? I think the most important reason is the compatiblity, documentation and knowledge share, is it right? so if we follow this principle,  it's better to continue the convention if some functions/structure are already accepted by multiple mature OS for many years even they haven't been standardized yet, because:
1.These are most likely the base for the next standard revision and then become the standard.
2.Improve the compatibility since many software already depends on this well established convention.
3.It's also consistent with some function/structure we already borrowed from Linux.
So I think the standard should be extended to cover the defacto standard(well established convention) too.

**Comment by patacongo** [comment]
I think we should follow the coding standard, but I have no problem with incorporating the PR as it is.  Feel free to merge it.

**Comment by patacongo** [comment]
I don't think the "making things easier" is a argument for not following the coding style.  That is spelled out in the Inviolables: 

     53   o Personal or organizational preference is not a justification for a
     54     coding style change.
     55   o Nothing can come into NuttX that does not follow the coding standard.
     56   o Expediency is not a justification for violating the coding standard.

**Comment by patacongo** [comment]
We all agreed to support these Inviolable principles when we started this project together.

**Comment by masayuki2009** [comment]
I've modified the code and pushed with -f.

And tested with the following environments
    (32bit cpu) spresense:elf, spresense:posix_spawn, spresense:module
    (64bit cpu) maix-bit:elf, maix-bit:posix_spawn, maix-bit:module (newly added)
